### PR TITLE
feat: unwrap casts of string and dictionary columns

### PR DESCRIPTION
**Pull request by erratic-pattern**
## Which issue does this PR close?
Closes https://github.com/apache/datafusion/issues/10220

## Rationale for this change

Improves performance of queries that filter string/dictionary columns against literals (`WHERE dict_col = 1`) by avoiding casting the column

## What changes are included in this PR?
Allows the optimizer to unwrap casts of string types in comparison operations. 

## Are these changes tested?
Yes 

## Are there any user-facing changes?
No

### CODE REVIEW

**Comment by erratic-pattern** [comment]
cc @jayzhan211  this one doesn't interfere with `comparison_cast`  unlike https://github.com/apache/datafusion/pull/10221

File: datafusion/sqllogictest/test_files/dictionary.slt [hunk 1]
```diff
@@ -386,3 +386,38 @@ drop table m3;
...
+02)--TableScan: test projection=[column1, column2]
+physical_plan
+01)CoalesceBatchesExec: target_batch_size=8192
+02)--FilterExec: column2@1 = 1
+03)----MemoryExec: partitions=1, partition_sizes=[1]
+
+
+# Now query using an integer which must be coerced into a dictionary string
+query T?
+SELECT * from test where column2 = 1;
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/dictionary.slt:420]: `explain` query for this is helpful too.

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 2]
```diff
@@ -298,19 +306,46 @@ fn is_support_data_type(data_type: &DataType) -> bool {
     )
 }
 
+/// Returns true if [UnwrapCastExprRewriter] supports casting this value as a string
+fn is_supported_string_type(data_type: &DataType) -> bool {
+    matches!(data_type, DataType::Utf8)
```
Comment (jayzhan211) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:311]: we can also support LargeUtf8.

**Review by jayzhan211 — APPROVED** [review]
👍

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 3]
```diff
@@ -298,19 +306,47 @@ fn is_support_data_type(data_type: &DataType) -> bool {
...
-    // the rule just support the signed numeric data type now
-    if !is_support_data_type(&lit_data_type) || !is_support_data_type(target_type) {
-        return Ok(None);
+    if !is_supported_type(&lit_data_type) || !is_supported_type(target_type) {
+        return None;
     }
     if lit_value.is_null() {
         // null value can be cast to any type of null value
-        return Ok(Some(ScalarValue::try_from(target_type)?));
+        return ScalarValue::try_from(target_type).ok();
```
Comment (alamb) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:331]: I wonder if ignoring errors will make tracking down unsupported types harder (I am imagining when we try and add REEArray or StringViewArray).

**Review by alamb — APPROVED** [review]
Thank you @erratic-pattern  and @jayzhan211  -- this is a really nice PR. Thank you for proceeding and pushing to get the "right" solution (rather than changing coercion)

**Comment by alamb** [comment]
I made a PR with a few more tests here: https://github.com/apache/datafusion/pull/10335

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs (hunk 3 above)
Comment (erratic-pattern) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:331]: > I wonder if ignoring errors will make tracking down unsupported types harder (I am imagining when we try and add REEArray or StringViewArray). 🤔

yeah we may want to reintroduce a Result type, but as it stands currently the errors were already being ignored anyway so they didn't serve any purpose.

also the boolean is_supported_* means that we'll just silently skip those cases anyway.

so realistically most of those errors just don't ever occur and if they did occur we wouldn't know about them. we would need to rethink the design of this code to surface those errors. 

another interesting possibility is debug assertions that panic, but only on 'debug' builds. I've always found this to be an interesting but underutilized tool for uncovering bugs

Comment (erratic-pattern) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:331]: see https://doc.rust-lang.org/std/macro.debug_assert.html

Comment (erratic-pattern) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:331]: Actually I was mistaken, these errors weren't being ignored. I could add them back in in a new PR?

Comment (alamb) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:331]: > Actually I was mistaken, these errors weren't being ignored. I could add them back in in a new PR?

I think that would be really nice if you had the time.

**Pull request merged**